# Introduction to Python Programming for Scientific Applications
### Day 1: The Basics




1. Variables for numbers, lists, and arrays
2. Importing and using modules; math, NumPy ++
3. while- loops and for- loops
4. functions
5. if-tests
6. plotting with matplotlib

## Installing Python

### A Quick Introduction to Jupyter Notebooks

This document, and most of the others in IN1910, are written in Jupyter notebooks. This is a tool that allows for combining traditional text with code that is executable. Previously known as IPython notebooks, the name was changed to reflect that it works with many different languages. Jupyter notebooks has grown into a popular tool for scientific programming, both for teaching and working. 

You are not required to become proficient at using or writing notebooks in this course, and you are free to simply read the documents, and do all your own coding in your preferred text editor. However, Jupyter is a nice tool to know, so also feel free to use it to solve exercises and play around with code and concepts.

In notebooks, everything consists of cells, which are either markdown (text, math, html, etc) or code. In the markdown cells you can render mathematics with LaTeX syntax, and you can include markdown tables, lists, and use html to include images. In the code cells you can write code, and then execute the cell to see the output. You can do this either by clicking the `Run` command in the toolbar on the top, or use the more efficient shortcut `Ctrl+Enter` or `Shift+Enter`. When you run a code, the output will appear below the cell. If you make a change to a cell that has already been run you can rerun the cell, and the new output will replace the old one.

One element that might be confusing to those that have not used notebooks before, is that the notebook is one interactive session, and so variables are remembered after running a cell. However, when loading up a notebook for the first time, everything is forgotten, and you will need to run the cells in order again for everything to work properly. Here a helpful tip is to use the `Kernel` > `Restart and Run All` functionality in the toolbar, which will run the whole notebook in order and load everything into memory.

### Variables and Types


In Python, as in most languages, we use `=` for assigning variables. We do not need to declare the type of the variables, this is implicit and decided by context

In [1]:
t = 0.5
v0 = 2
a = 0.2
s = v0*t + 0.5*a*t**2

Here we first assign the three variables `t`, `v0` and `a`. Then we assign the variable `s` based on our other variables by doing some arithmetic. An assignment on the form of the final line is interpreted as follows:
* Evaluate the right-hand side of the `=` symbol, this results in an *object*
* The left hand side is a name for that object
In this case, the result of the computation is a number, and `s` becomes a name for that number. But how do we check what `s` is? We use the `print`-statement.

In [2]:
print(s)

1.025


#### Controlling how you print

When printing out information, we can do so in different ways

https://docs.python.org/3/tutorial/inputoutput.html

* Bankers' rounding and FLOATING precision


In [3]:
print("After {:.2f} seconds, the object has traveled {:.3f} meters".format(t, s))

After 0.50 seconds, the object has traveled 1.025 meters


Here we define a text string to be written out by using `"`, inside the text string we define replacement fields by writing `{}` and then we use `.format` to fill these in with given values. Inside the replacement fields we write `:.2f` and `:.3f` to write precisely 2 and 3 decimals respectively. (The `f` stands for *float*)

Note that if print out the distance with two decimals, we get a bit of a surprising result:

In [5]:
print("{:.2f}".format(s))

1.02


1.025 is rounded to 1.02 and not to 1.03! This is because we are running Python3, which uses something called [*bankers' rounding*](https://en.wikipedia.org/wiki/Rounding), in which a .5 is rounded down if the preceeding number is even, and up if it is odd.

In [7]:
print(round(0.5))
print(round(1.5))
print(round(2.5))
print(round(3.5))

0
2
2
4


#### Importing 

Python includes a large number of standard libraries and external packages. In IN1910 we will for example make use of the SciPy stack, which includes packages like numpy, scipy, matplotlib, etc. We can import packages in several ways

In [8]:
from math import *
import scipy
import numpy as np
from scipy.integrate import solve_ivp

1. These are different examples of importing. The first line uses a *wildcard*-import to import everything in math. Note that while this is OK for a quick example, it is considered bad practice and should be avoided. 
2. The second statement imports scipy as a package, so now we can use it by writing for example: `scipy.e` The preference for doing this over a wildcard-import is that it keeps the scipy namespace, meaning we (1) don't overwrite anything by accident, and (2) anywhere we use anything from scipy, it will be very apparent from the code.
3. The third import is very similar to the second, but we just rename the namespace from numpy to np. This is very common and is just to save some typing later on, as we often use numpy functions quite a lot in our programs. Note that "np" is a convention, and you should use exactly that, and not something else. Same for matplotlib.pyplot, which is often shortened to "plt".
4. The fourth import imports a single function from a package. If you know ahead of time you will only use this single feature from the package. It's a nice way to make that apparent to the user.

Import statements should always appear at the top of modules. Modules are what we call seperate python-files.